### PR TITLE
linked hash set, linked hash map

### PR DESCRIPTION
**Pull request by akiezun**
using linked or sorted data structures to reduce our dependency on the arbitrary changes in iterator orders (happens everytime you move from a jvm to another jvm)

### CODE REVIEW

**Comment by akiezun** [comment]
I'm going to abandon this - Picard tests depend on the order in contrived ways, eg:
```
    private DetailPair getWorstMetrics(final List<DetailPair> metrics) {
        PreAdapterDetailMetrics worstPreAdapterMetrics = null;
        BaitBiasDetailMetrics worstBaitBiasMetrics = null;
        for (final DetailPair m : metrics) {
            if (worstPreAdapterMetrics == null || m.preAdapterMetrics.QSCORE < worstPreAdapterMetrics.QSCORE) worstPreAdapterMetrics = m.preAdapterMetrics;
            if (worstBaitBiasMetrics == null || m.baitBiasMetrics.QSCORE < worstBaitBiasMetrics.QSCORE) worstBaitBiasMetrics = m.baitBiasMetrics;
        }
        return new DetailPair(worstPreAdapterMetrics, worstBaitBiasMetrics);
    }
```

if all `QSCORE` are the same this arbitrarily pick the first one. That backfires when I switch to linked hashmaps and linkedhashsets. Specifically `CollectSequencingArtifactMetricsTest` fails. Given that #1210 (fate of Picard) is unresolved I'm going to not put time into it yet.

**Comment by droazen** [comment]
How about doing this everywhere except in Picard, instead of abandoning it completely? Non-linked hash sets/maps have caused MANY problems in GATK over the years.

**Comment by ronlevine** [comment]
The issue is that only the linked list version of hash sets/maps do has a predictable iteration order. So, any algorithm that is order dependent may get different results between different Java versions. The only downside is the linked list will take up some additional storage.

**Comment by akiezun** [comment]
fine but someone else does it when picard is decided one way or the other

**Comment by akiezun** [comment]
@droazen can you review?

File: src/main/java/org/broadinstitute/hellbender/tools/ClipReads.java [hunk 1]
```diff
@@ -505,7 +505,7 @@ public SeqToClip(String name, byte[] bytez) {
         public long nRangeClippedBases = 0;
         public long nSeqClippedBases = 0;
 
-        HashMap<String, Long> seqClipCounts = new HashMap<>();
+        SortedMap<String, Long> seqClipCounts = new TreeMap<>();
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/ClipReads.java]: Why did you change `HashMap` to `TreeMap` and not `LinkedHashMap` here?

File: src/main/java/org/broadinstitute/hellbender/tools/spark/validation/CompareDuplicatesSpark.java [hunk 2]
```diff
@@ -106,7 +105,7 @@ protected void runTool(final JavaSparkContext ctx) {
         }
         if (printSummary) {
             MatchType[] values = MatchType.values();
-            HashSet<MatchType> matchTypes = Sets.newHashSet(values);
+            Set<MatchType> matchTypes = Sets.newHashSet(values);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/validation/CompareDuplicatesSpark.java]: Does this produce a `LinkedHashSet`? If not, should you manually construct one?

File: src/test/java/org/broadinstitute/hellbender/utils/collections/CountSetUnitTest.java [hunk 3]
```diff
@@ -30,7 +27,7 @@ public void testSize(final int capacity) {
     public void testSingleValueAdd() {
         final int CAPACITY = 10;
         final CountSet subject = new CountSet(CAPACITY);
-        final HashSet<Integer> reasuranceSet = new HashSet<>(CAPACITY);
+        final HashSet<Integer> reasuranceSet = new LinkedHashSet<>(CAPACITY);
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/collections/CountSetUnitTest.java]: type on the left side should be `Map`

File: src/test/java/org/broadinstitute/hellbender/utils/collections/CountSetUnitTest.java [hunk 4]
```diff
@@ -118,7 +115,7 @@ public void testMinMax() {
     public void testIncrease() {
         final int CAPACITY = 10;
         final CountSet subject = new CountSet(CAPACITY);
-        final HashSet<Integer> reasuranceSet = new HashSet<>(CAPACITY);
+        final HashSet<Integer> reasuranceSet = new LinkedHashSet<>(CAPACITY);
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/collections/CountSetUnitTest.java]: type on the left side should be `Set`

File: src/test/java/org/broadinstitute/hellbender/utils/collections/NestedIntegerArrayUnitTest.java [hunk 5]
```diff
@@ -57,7 +57,7 @@ public void testPutget1Key() throws Exception {
         arr.put("bozo", 1);
 
         Assert.assertEquals(2, arr.getAllLeaves().size());
-        Assert.assertEquals(Sets.newHashSet("fred", "bozo"), new HashSet<>(arr.getAllValues()));
+        Assert.assertEquals(Sets.newHashSet("fred", "bozo"), new LinkedHashSet<>(arr.getAllValues()));
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/collections/NestedIntegerArrayUnitTest.java]: Can you audit the codebase for uses of `Sets.newHashSet()`, and determine whether they'll cause any issues (eg., by creating non-linked `HashSets`)

File: src/test/resources/org/broadinstitute/hellbender/tools/ClipReads/expected.clippingReadsTest.withRG.hg19.QT_10_CT_15_X_CCCCC_XF.tmp [hunk 6]
```diff
@@ -8,7 +8,7 @@ Percent of clipped bases              24.70
 Number of quality-score clipped bases 126
 Number of range clipped bases         65
 Number of sequence clipped bases      53
-         8 clip sites matching GAGG
         10 clip sites matching CCCCC
+         8 clip sites matching GAGG
```
Comment (droazen) [inline src/test/resources/org/broadinstitute/hellbender/tools/ClipReads/expected.clippingReadsTest.withRG.hg19.QT_10_CT_15_X_CCCCC_XF.tmp]: Is this the only change in test output across all tools?

**Comment by droazen** [comment]
Review complete, back to @akiezun, merge after addressing comments.

File: src/main/java/org/broadinstitute/hellbender/tools/ClipReads.java (hunk 1 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/ClipReads.java]: the `seqClipCounts` are better organized when sorted by the String key than just in arbitrary order. So sorted map is better.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/validation/CompareDuplicatesSpark.java (hunk 2 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/validation/CompareDuplicatesSpark.java]: good catch. No, it does not produce a linkedhashset. I changed it in a few places where appropriate

File: src/test/java/org/broadinstitute/hellbender/utils/collections/CountSetUnitTest.java (hunk 3 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/collections/CountSetUnitTest.java]: fixed here and throughout the codebase

File: src/test/java/org/broadinstitute/hellbender/utils/collections/CountSetUnitTest.java (hunk 4 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/collections/CountSetUnitTest.java]: fixed

File: src/test/java/org/broadinstitute/hellbender/utils/collections/NestedIntegerArrayUnitTest.java (hunk 5 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/collections/NestedIntegerArrayUnitTest.java]: yes, all done

File: src/test/resources/org/broadinstitute/hellbender/tools/ClipReads/expected.clippingReadsTest.withRG.hg19.QT_10_CT_15_X_CCCCC_XF.tmp (hunk 6 above)
Comment (akiezun) [inline src/test/resources/org/broadinstitute/hellbender/tools/ClipReads/expected.clippingReadsTest.withRG.hg19.QT_10_CT_15_X_CCCCC_XF.tmp]: yes, the only change was to resort those guys (which makes more sense then list them in an arbitrary order)
